compiler/circle-mpqsolver/src/bisection/ErrorApproximator.cpp

```diff
@@ -0,0 +1,386 @@
+    mult *= (stride->h() * stride->w());
+  }
+  else if (node->opcode() == CircleOpcode::TRANSPOSE_CONV)
+  {
+    auto nconv = loco::must_cast<const CircleTransposeConv *>(node);
+    auto stride = nconv->stride();
+    mult *= (stride->h() * stride->w());
+  }
+
+  return mult;
+}
+
+uint32_t get_computations_per_channel(const CircleNode *node)
+{
+  uint32_t computations_per_channel = 1;
+  std::vector<uint32_t> cir_shape;
+  // size of activation map
+  auto activation = get_activation_node(node);
+  assert(activation != nullptr);
+  if (get_shape(activation, cir_shape))
+  {
+    for (uint32_t index = 0; index < cir_shape.size(); index++)
+    {
+      computations_per_channel *= cir_shape[index];
+    }
+  }
+
+  if (auto weights = loco::must_cast<CircleNode *>(get_weight(node)))
+  {
+    std::vector<uint32_t> w_shape;
+    if (get_shape(weights, w_shape))
+    {
+      computations_per_channel *= (w_shape[1] * w_shape[2]); //[channels_out, k_x, k_y, channels_in]
+    }
+  }
+
+  uint32_t stride = get_stride_multiplier(node);
+  if (stride < 1)
+  {
+    throw std::runtime_error("Failed to find correct stride index in " + node->name());
+  }
+  return computations_per_channel / stride;
+}
+
+void get_min_max_activation_values(const CircleNode *node, float &ci_min, float &ci_max)
+{
+  auto preds = loco::preds(node);
+  for (const auto &pred : preds)
+  {
+    auto parent_node = loco::must_cast<const luci::CircleNode *>(pred);
+    if (has_min_max(parent_node))
+    {
+      auto quantparam = parent_node->quantparam();
+      if (quantparam->min.size() > 0)
+      {
+        ci_min = quantparam->min[0];
+        ci_max = quantparam->max[0];
+      }
+    }
+  }
+}
+
+} // namespace
+
+namespace mpqsolver
+{
+namespace bisection
+{
+
+/**
+ * How Approximate works?
+ *
+ * Currently it works just for convolution layers, but may be generalized for other types as well.
+ * See discussion at https://github.com/Samsung/ONE/pull/10170#discussion_r1042246598
+ * Convolution is just a matrix multiplication.
+ * While quantizing we introduce quantization error into convolution operand (activations) as well
+ * as into convolution weights. A_q * W_q = (A + q_err(A)) * (W + q_err(W)) = A * W + A * q_err(W) +
+ * W * q_err(A), q_err - quantization error, W - weight matrix, A - activations from previous layer,
+ * so quantization error of matrix multiplication is A * q_err(W) + W * q_err(A).
+ * Estimating its largest value we get
+ * ~ number_of_additions * (A_max * (W_max - W_min) / 255 + W_max * (A_max - A_min) / 255)
+ * The following code tries to get total error for quantizing convolution node into Q8.
+ * It's just an heuristic (Metric sensitivity depends highly on derivatives as well).
+ */
+float approximate(const CircleNode *node)
+{
+  auto opcode = node->opcode();
+  if (opcode != CircleOpcode::CONV_2D && opcode != CircleOpcode::TRANSPOSE_CONV &&
+      opcode != CircleOpcode::DEPTHWISE_CONV_2D)
+  {
+    return 0.0; // TODO
```

Comment:
Are you going to implement error functions for other operators? If so, please add one more step, e.g., use switch statement in this function and call `approximate_conv` if opcode is CONV_2D/TRANSPOSE_CONV/DEPTHWISE_CONV_2D. Return 0 for the default case. This will make easier to implement error functions for other ops.

Reply:
Or, you can use visitor pattern to implement error function operator-by-operator.